text/0104-string-parameterization-output-spec.md

```diff
@@ -0,0 +1,264 @@
+
+```sql
+SELECT * FROM cities WHERE country = 'CA';
+```
+
+```sql
+SELECT * FROM cities WHERE country = 'US';
+```
+
+```sql
+SELECT * FROM cities WHERE country = 'AU';
+```
+
+A successful parameterizer will ingest these span descriptions, and determine that `"CA"`, `"US"`, and `"AU"` were probably a parameter inserted dynamically by the application depending on some condition. The three query descriptions are examples of this parameterized query:
+
+```sql
+SELECT * FROM cities WHERE county = ?;
+```
+
+where `?` is a country code parameter that varies from query to query.
+
+This spec outlines standards for the output of a parameterizer, including but not limited to the desired string formatting, payload structure, and character escaping. The spec is organized by data type. Parameterization methods and infrastructure are outside the scope of this spec.
+
+## Motivation
+
+The main goal of parameterization is to remove dynamic values from strings, in order to group “the same” strings together. “The same” loosely means that the strings represent the same underlying resource. For example, the strings `GET /users/17` and `GET /users/781` might represent accessing the web application route `/users/:id`. The strings `SCARD org:18:active` and `SCARD org:98:active` represent access to the same kind of Redis resource, and so on. When looking at application data (traces, URLs, spans, etc.) it is most useful to look at parameterized values, because they represent aggregate behaviour. In the routing example, the `/users/:id` route is a useful entity to examine, while the unparameterized URLs are not. Parameterization enables aggregation, and aggregation makes it possible to store, query, and compare data easily.
+
+Sentry performs parameterization in many different part of the application. The SDKs, Relay, and Ingest are three examples. There is no agreement between these parts beyond conventions. As a result, the output of parameterization is inconsistent.
+
+Here are three ways different SDKs might parameterize a SQL query:
+
+- `SELECT * FROM countries WHERE id = ?`
+- `SELECT * FROM countries WHERE id = %s`
+- `SELECT * FROM countries WHERE id = $1`
+
+As another example, the SDKs might pass the query `SELECT * FROM country WHERE code IN (?, ?)` to Ingest, which will further parameterize that as `SELECT * FROM country WHERE code IN (?)`. The string `(?, ?)` is technically only partially parameterized from the system's perspective.
+
+This has important downstream effects on user experience. A good parameterizer will correctly group similar strings together in a way that represents the original intent. If parameterization is successful, users will see groups that make sense to them. Good parameterization will also reduce cardinality, and therefore reduce data storage costs. A bad parameterizer will fill the users’s lists of transactions and spans with noisy similar descriptions that drown out the signal.
```

Comment:
Everything in this paragraph is true, but how would standardizing the placeholder format help with this?

Standardizing could be a good idea, but this section could use some more convincing points as to why that is IMO. Do we have examples of cases where the product fails to deal with placeholders because they are in the wrong format, or a string is altered by two different parameterizers and therefore looks messy (e.g. `/{my}/<parameterized>/:string`?).

Reply:
Great feedback, thank you! I will update this section. There are actually a few goals here, and I didn't explain them fully.

1. Downstream processing edge cases. When SDKs use different placeholder formats, all downstream processing of descriptions has to handle each kind of placeholder. We ran into this with performance issues, suspect spans, and now span metrics. All three of them have to handle `?` vs. `%s` vs. `$1` as database placeholders. If a new SDK comes in with a fourth format, all have to be updated
2. We haven't seen any UI cases of mixed parameterization, but that's what I'm trying to prevent. For example, transaction names on the frontend and backend are both URL-like, but look different. `/api/0/organizations/{organization_slug}/events-trends-stats/` vs. `/projects/:projectId/`. Why curly brackets in one, but colons in the other? Right now they're never mixed, but it's bad precedent

You're right, a standard placeholder does _not_ inherently help with cardinality. One aspect of the spec that _does_ help with cardinality is demanding that lists of arguments like `IN (? ? ? ?)` are collapsed. Otherwise they'd all have different descriptions. Not a very strong case, but it's one example!